core/src/smartcontracts/isi/block.rs

```diff
@@ -10,10 +13,43 @@ impl ValidQuery for FindAllBlocks {
     fn execute(&self, wsv: &WorldStateView) -> Result<Self::Output, query::Error> {
         let mut blocks: Vec<BlockValue> = wsv
             .blocks()
-            .map(|blk| blk.clone())
+            .map(|block| block.clone())
             .map(VersionedCommittedBlock::into_value)
             .collect();
         blocks.reverse();
         Ok(blocks)
     }
 }
+
+impl ValidQuery for FindAllBlockHeaders {
+    #[metrics(+"find_all_block_headers")]
+    fn execute(&self, wsv: &WorldStateView) -> Result<Self::Output, query::Error> {
+        let mut blocks: Vec<BlockHeaderValue> = wsv
+            .blocks()
+            .map(|block| block.clone())
+            .map(VersionedCommittedBlock::into_value)
+            .map(|block_value| block_value.header)
+            .collect();
+        blocks.reverse();
+        Ok(blocks)
+    }
+}
+
+impl ValidQuery for FindBlockHeaderByHash {
+    #[metrics(+"find_block_header")]
+    fn execute(&self, wsv: &WorldStateView) -> Result<Self::Output, query::Error> {
+        let hash = self
+            .hash
+            .evaluate(wsv, &Context::default())
+            .wrap_err("Failed to evaluate hash")
+            .map_err(|e| query::Error::Evaluate(e.to_string()))?
+            .typed();
+
+        let block = wsv
+            .blocks()
+            .find(|block| block.hash() == hash)
+            .ok_or_else(|| query::Error::Find(Box::new(FindError::Block(hash))))?;
```

Comment:
Hm, looks like very long operation but let's keep it now. We don't know real world scenarios and don't need preventual optimisations

Reply:
There's no faster way to find a block by hash anyway, we would need to rework how the `Chain` stores blocks to do it.

Reply:
Chain might not store all blocks in memory in the future. The solution has to involve a `MerkleTree`, which would allow you to find the object given hash in $O(\log n)$